_posts/2021-11-23-how-to-dry-your-ref-arch.adoc

```diff
@@ -0,0 +1,588 @@
+Terragrunt configuration and semantically compare the JSON outputs.
+
+. Install https://github.com/tmccombs/hcl2json[hcl2json] and https://github.com/josephburnett/jd[jd]. `hcl2json` will be used to convert the Terragrunt configurations to `json`, and `jd` will be used to create a semantic diff between the two.
+. For each Terragrunt configuration, convert the `terragrunt.hcl` file to `json` using `hcl2json`:
+** `hcl2json dev/us-west-2/dev/networking/vpc/terragrunt.hcl > dev/us-west-2/dev/networking/vpc/terragrunt.hcl.json`
+. For Terragrunt configuration, run through `jd` against the reference point to identify differences in the json file.
+** `jd -set dev/us-west-2/dev/networking/vpc/terragrunt.hcl.json stage/us-west-2/stage/networking/vpc/terragrunt.hcl.json`
+
+
+We use `-set` mode to make it easier to understand which blocks and attributes are different. In `set` mode, `jd` will output each diff in the following format:
+
+----
+@ [KEYS,TO,ITEM]
+- REMOVED
++ ADDED
+----
+
+Each element in the list after `@` indicates the index to the item that is different. For example, in the following, the
+first diff represents a difference in the `Title` attribute of the movie object that is in the 67th position of the
+`Movies` list:
+
+----
+@ ["Movies",67,"Title"]
+- "Dr. Strangelove"
++ "Dr. Evil Love"
+@ ["Movies",67,"Actors","Dr. Strangelove"]
+- "Peter Sellers"
++ "Mike Myers"
+@ ["Movies",102]
++ {"Title":"Austin Powers","Actors":{"Austin Powers":"Mike Myers"}}
+----
+
+Like with `diff`, cross off any blocks, attributes, and input keys from the initial list that are different based
+on the output. In the `jd` output, this would be the first element in each `@` entry, or the second element of each `@`
+list that has `input` as the first element.
+
+
+[[step_3]]
+==== Step 3: Extract common configurations
+
+Once you have identified the list of common blocks, attributes, and input keys, the next step is to create a common
+Terragrunt configuration that includes these values.
+
+. Create a new file to house the common configuration. This should be placed somewhere that can be easily linked to
+from the root of the repository. We recommend using the following folder structure:
+
++
+----
+.
+├── terragrunt.hcl
+└── _envcommon
+    └── CATEGORY
+        └── RESOURCE.hcl
+----
+
++
+Where the common configuration files are placed in `_envcommon/CATEGORY/RESOURCE.hcl`. `CATEGORY` and `RESOURCE` should
+mimic the base folder structure of the Reference Architecture:
+
++
+----
+.
+└── ACCOUNT
+    ├── REGION
+    │   ├── ENVIRONMENT
+    │   │   └── CATEGORY
+    │   │       └── RESOURCE
+    │   └── _regional
+    │       └── RESOURCE
+    └── _global
+        └── RESOURCE
+----
+
++
+For example, for the `vpc` component, the `CATEGORY` will be `networking`, and the `RESOURCE` will be `vpc`, resulting
+in a common configuration file located at `_envcommon/networking/vpc.hcl`.
```

Comment:
The directory trees are gold! And your example here to illustrate that is very helpful.